attalos/imgtxt_algorithms/approaches/negsampling.py

```diff
@@ -98,32 +97,32 @@ def _get_ids(self, tag_ids, numSamps=[5, 10], uniform_sampling=False):
         """
         Takes a batch worth of text tags and returns positive/negative ids
         """
-        num_images_in_batch = len(tag_ids)
-        num_positive_samples = numSamps[0]
-        num_negative_samples = numSamps[1]
-        pos_word_ids = np.ones((num_images_in_batch, num_positive_samples), dtype=np.int32)*-1
+        pos_word_ids = np.ones((len(tag_ids), numSamps[0]), dtype=np.int32)
+        pos_word_ids.fill(-1)
         for ind, tags in enumerate(tag_ids):
             if len(tags) > 0:
-                # If there are any valid tags choose 1
-                num_tags_for_image = len(tags)
-                pos_word_ids[ind] = np.random.choice(tags, size=num_positive_samples)
+                pos_word_ids[ind] = np.random.choice(tags, size=numSamps[0])
         
         neg_word_ids = None
         if uniform_sampling:
             neg_word_ids = np.random.randint(0, 
                                              self.one_hot.vocab_size, 
-                                             size=(len(tag_ids), num_negative_samples))
+                                             size=(len(tag_ids), numSamps[1]))
         else:
-            neg_word_ids = np.ones((num_images_in_batch, num_negative_samples), dtype=np.int32)*-1
-            for ind, tag_inds in enumerate(tag_ids):
-                neg_word_ids[ind] = self.negsampler.negsamp_ind(pos_word_ids[ind], num_negative_samples)
+            neg_word_ids = np.ones((len(tag_ids), numSamps[1]), dtype=np.int32)
+            neg_word_ids.fill(-1)
+            for ind in range(pos_word_ids.shape[0]):
+                # NOTE: This function call should definitely be pos_word_ids[ind]
+                #          but that results in significantly worse performance
+                #          I wish I understood why.
+                #          I think this means we won't sample any tags that appear in the batch
+                neg_word_ids[ind] = self.negsampler.negsamp_ind(pos_word_ids, numSamps[1])
```

Comment:
Agreed with the comment above. Passing in `pos_word_ids` as the `ignored_inds` for `negsamp_ind()` excludes every positive tag from the batch from being selected during negative sampling, whereas passing in `pos_word_ids[ind]` only excludes positive tags from the current image. Is this desired behavior? Maybe @karllab41 knows?